### PR TITLE
Use ipld-prime for encode & decode in DAG Service, construct ipld-prime compatible nodes

### PR DESCRIPTION
**Pull request by hannahhoward**
This is a minimal shim to put IPLD prime into IPFS.

It means that:
- all serialization / deserialization is done through IPLD prime
- all nodes returned by the dag service are in fact IPLD-prime compatible nodes
- the protobuf node implementation is still very WIP - we're bumping up against the fact that IPLD-prime makes the (I think very sane) decision to keep nodes themselves immutable, and force you to use builder interfaces for mutations, but the existing ProtoNode implementation contains a number of mutation methods.

### CODE REVIEW

File: coding.go [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
 
 	blocks "github.com/ipfs/go-block-format"
-	pb "github.com/ipfs/go-merkledag/pb"
-
 	cid "github.com/ipfs/go-cid"
-	ipld "github.com/ipfs/go-ipld-format"
+	format "github.com/ipfs/go-ipld-format"
+	pb "github.com/ipfs/go-merkledag/pb"
+	dagpb "github.com/ipld/go-ipld-prime-proto"
+	"github.com/ipld/go-ipld-prime/fluent"
```
Comment (willscott) [inline coding.go:14]: @warpfork - is fluent the right dialect of prime to be using here?

File: node.go [hunk 2]
```diff
@@ -382,3 +387,17 @@ func (n *ProtoNode) Tree(p string, depth int) []string {
 	}
 	return out
 }
+
+func ProtoNodeConverter(b blocks.Block, nd ipld.Node) (legacy.UniversalNode, error) {
+	pbNode, ok := nd.(dagpb.PBNode)
```
Comment (willscott) [inline node.go:392]: is there a todo to add here to do codec stuff if it isn't dagpb, rather than failing?

File: prime.go [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+// LookupByIndex is the equivalent of LookupByString but for indexing into a list.
+// As with LookupByString, the returned Node may be any of the Kind:
+// a primitive (string, int64, etc), a map, a list, or a link.
+//
+// If the Kind of this Node is not Kind_List, a nil node and an error
+// will be returned.
+//
+// If idx is out of range, a nil node and an error will be returned.
+func (n *ProtoNode) LookupByIndex(idx int64) (ipld.Node, error) {
+	_, err := n.EncodeProtobuf(false)
```
Comment (willscott) [inline prime.go:60]: hopefully this is cached? is it preferable to have a `n.GetEncoded` that provides the pointer to the internal node and computes it if it doesn't exist?

**Review by willscott — APPROVED** [review]
🎉

File: coding.go (hunk 1 above)
Comment (hannahhoward) [reply coding.go:14]: I suspect it is -- fluent is basically the way to do a bunch of stuff without an incredibly verbose syntax.

File: coding.go [hunk 4]
```diff
@@ -22,38 +25,77 @@ const _ = pb.DoNotUpgradeFileEverItWillChangeYourHashes
...
-	n.encoded = encoded
-	return nil
+	n.links = links
+	return n
+}
+func (n *ProtoNode) marshalImmutable() (*immutableProtoNode, error) {
+	nb := dagpb.Type.PBNode.NewBuilder()
+	err := fluent.Recover(func() {
+		fb := fluent.WrapAssembler(nb)
+		fb.CreateMap(-1, func(fmb fluent.MapAssembler) {
```
[the remark this answers is not in the document]
Comment (hannahhoward) [reply coding.go:60]: I guess that's the case? I honestly copied this real quick from the go-ipld-prime-proto code and I can't remember why I used -1 here.

File: coding.go (hunk 1 above)
Comment (warpfork) [reply coding.go:14]: You can also try either [`quip`](https://pkg.go.dev/github.com/ipld/go-ipld-prime/fluent/quip) or [`qp`](https://pkg.go.dev/github.com/ipld/go-ipld-prime/fluent/qp) dialects, if you like -- they're questing for good syntax but both should have a lower performance overhead than the `fluent` package does.

`quip` and `qp` are more recently introduced than `fluent`, and someday we'll probably try to narrow down how many of these there are, but for now all three are considered valid (and battle-testing so we can get hands-on feedback about the ergonomics is highly welcome :)).

Comment (willscott) [reply coding.go:14]: I think that was my question: should we prefer quip / qp vs fluent for this dev?

File: prime.go (hunk 3 above)
Comment (hannahhoward) [reply prime.go:60]: that's actually what the somewhat inaccurately named EncodeProtobuf actually does -- the boolean paramenter, assuming it's false, means don't re-encode unless the data has changed.

File: node.go (hunk 2 above)
Comment (hannahhoward) [reply node.go:392]: other codecs should "just work" based on the design of go-ipld-prime. Protobuf & Raw are special cases.

File: coding.go (hunk 1 above)
Comment (hannahhoward) [reply coding.go:14]: AND I was unaware of quip/qp so now I need ot check those out :)

**Comment by hannahhoward** [comment]
superseded by #67

**Pull request closed without merging**